### ISSUE TITLE
Encoding issues with codekit's server

### ISSUE THREAD

**Issue by inseldesign** [bdkjones/CodeKit#360]
**Quick, short summary:**

Since last update of codekitapp (v3.1) I have some issues with nouns encoding. Ä and Ö and Ü are not encoded in preview. So it ends up with something like "Ã¶"

I tested in chrome, safari and brave. So i guess it should be something codekit internal.

Thanks :-)

**Comment by bdkjones** [comment]
Which character set are you using on the page? It must be UTF-8. Likewise, which text-encoding do you have your browsers set to use? Again, it must be UTF-8.

Please post a small demo project showing the issue.

**Comment by inseldesign** [comment]
Hi bryan,

I use UTF-8 (like the other years and versions of codekit before). I started with a normal HTML file set the meta charset to UTF 8 (like every time, nothing special).

I also checked the html-file directly (by double clicking) and this is working as expected (with all nouns encoded the right way).

**Comment by bdkjones** [comment]
Please post the demo project showing the issue.

**Comment by inseldesign** [comment]
ok. just take this code (https://codepen.io/slichtenstein/pen/aWqowP), and compile it via codekit. than preview it.
![05_09_17 at 07 08 57pm](https://cloud.githubusercontent.com/assets/5708900/25863196/25843b6c-34eb-11e7-9a04-3e32fd81e2c6.png)

**Comment by bdkjones** [comment]
Right. The reason I ask for a demo project is so that I can see what settings you're using, etc. It's not enough to just hand me a string because my default Pug settings may be different than yours and I won't reproduce the same result over here and then we'll both waste a lot of time writing things like this where we try to understand who's using what language and settings. (For example, your original post made ZERO mention of Pug, which is obviously the issue here.)

Hence the demo project request. I've been doing this for a LONG time and there really *IS* a method to my madness, I promise.

**Comment by inseldesign** [comment]
OK, Bryan, here we go. I made a simple project with no PUG, only a simple index.html, I got the same problem. Hopefully this helps to figure out whats going wrong. 

[Archiv.zip](https://github.com/bdkjones/CodeKit/files/988697/Archiv.zip)

**Comment by inseldesign** [comment]
ok got it. sorry. i never will code again :(

**Comment by bdkjones** [comment]
Wait, what? That's not the idea. I'm able to reproduce the issue with your project. Did you find a way to fix it?

**Comment by inseldesign** [comment]
I just   made an mistake with spaces in  the pug file. and guess  what: it was not the first time :-(

**Comment by inseldesign** [comment]
MEA CULPA!

**Comment by bdkjones** [comment]
Well, there's no pug files in the demo project and I'm still showing the incorrect characters. Can you be more specific and walk me through the resolution?

**Comment by inseldesign** [comment]
just check the head. even in html file it is closing without the meta tags inside. codekit works well (like ever).

**Comment by inseldesign** [comment]
But nevertheless the browser accepted this kind of html code when opening the file directly.

**Comment by bdkjones** [comment]
Ah, yea, I do see that. It's curious that the page renders differently when previewed through CodeKit versus when pulled up directly in Safari. My *guess* would be that the HTTP headers are different and in the latter case there's one that's telling the browser to render as UTF-8. CodeKit strips and modifies some HTTP headers to eliminate things like caching, etc. I'd have to look deeper, but I'd expect that to be the reason.

**Comment by bdkjones** [comment]
Oh God. Was this ever a rabbit hole.

I generally compare the result of CodeKit's internal HTTP server to what MAMP delivers, because MAMP runs standard Apache, which has always served as my benchmark. It turns out that if you fire up MAMP and host your demo project, the page is displayed with the correct characters.

Digging deeper, there was *no* difference in HTTP request/response headers between the MAMP-hosted page and the page hosted by CodeKit's internal server.

*HOWEVER*, CodeKit's internal server parses the string (the text of the page) and adds its special JavaScript snippet to enable browser-refreshing. The app then converts the string to *data* and passes that to the browser. This conversion is done explicitly with UTF-8 encoding. That is, the actual string contents of the page are converted to binary data in a UTF-8 (variable character byte length) format. 

That blob of data is then sent to the browser. This is where the trouble is. CodeKit does NOT currently add a `charset=utf-8` entry to the `Content-Type` HTTP header to inform the browser that the blob of data it's receiving is UTF-8 encoded (no matter what any `meta` tag on the page may say). So my best guess is that the browser is defaulting to `ISO-fucking-something-or-other` for the character encoding, instead of UTF-8.

Because CodeKit's internal server will ALWAYS use UTF-8, the app should add/modify the `charset` property of the `Content-Type` HTTP header to inform the browser that's the case. MAMP must be encoding the data in some different way that automatically triggers the browser to use the correct charset. 

Long story short, I've now changed things so that when you're using CodeKit's internal server, the `charset` HTTP header property is always set to `utf-8`, which forces the browser to interpret the data it has received in the correct format.

Because CodeKit also operates as a reverse-proxy server if you enable the "External Server" option in Project Settings, the app will also parse data received from an external origin server (like MAMP) and then re-encode that text as UTF-8 data. (Again, to add the special JS snippet that enables browser reloading.) Because of this, even if you're using an external server, CodeKit will still modify/add the `charset` HTTP header property.

This is why your demo project worked when you corrected your typo. As soon as the browser saw the `meta` tag specifying UTF-8, it rendered the page correctly. However, when that tag was missing (because of the typo), it had no `charset` HTTP header to fall back on and defaulted to the wrong text encoding. 

As a byproduct of this, I've also marginally sped up the proxy server when you have a project that uses an External Server.

TL;DR: Unicode is a goddamn nightmare.

**Comment by inseldesign** [comment]
Thanks for your extensively explanation. Thanks for all your hard work. Thanks for CODEKIT 👍
